### PR TITLE
chore: cache dns resolution in node

### PR DESCRIPTION
**Pull request by Jayachand**
## What are the changes introduced in this PR?

Write a brief explainer on your code changes.

## What is the related Linear task?

fixes DAT-466

## Please explain the objectives of your changes below

Put down any required details on the broader aspect of your changes. If there are any dependent changes, **mandatorily** mention them here

### Any changes to existing capabilities/behaviour, mention the reason & what are the changes ?

N/A

### Any new dependencies introduced with this change?

N/A

### Any new generic utility introduced or modified. Please explain the changes.

N/A

### Any technical or performance related pointers to consider with the change?

N/A

@coderabbitai review

<hr>

### Developer checklist

- [ ] My code follows the style guidelines of this project

- [ ] **No breaking changes are being introduced.**

- [ ] All related docs linked with the PR?

- [ ] All changes manually tested?

- [ ] Any documentation changes needed with this change?

- [ ] Is the PR limited to 10 file changes?

- [ ] Is the PR limited to one linear task?

- [ ] Are relevant unit and component test-cases added in **new readability format**?

### Reviewer checklist

- [ ] Is the type of change in the PR title appropriate as per the changes?

- [ ] Verified that there are no credentials or confidential data exposed with the changes.

### CODE REVIEW

**Comment by devops-github-rudderstack** [comment]
Test report for this run is available at: https://test-integrations-dev.s3.amazonaws.com/integrations-test-reports/rudder-transformer/3495/test-report.html

File: src/util/utils.js [hunk 1]
```diff
@@ -15,37 +16,61 @@ const BLOCK_HOST_NAMES_LIST = BLOCK_HOST_NAMES.split(',');
 const LOCAL_HOST_NAMES_LIST = ['localhost', '127.0.0.1', '[::]', '[::1]'];
 const LOCALHOST_OCTET = '127.';
 const RECORD_TYPE_A = 4; // ipv4
+const DNS_CACHE_ENABLED = process.env.DNS_CACHE_ENABLED === 'true';
+const DNS_CACHE_TTL = process.env.DNS_CACHE_TTL ? parseInt(process.env.DNS_CACHE_TTL, 10) : 300;
+const dnsCache = new NodeCache({
+  useClones: false,
+  stdTTL: DNS_CACHE_TTL,
+  checkperiod: DNS_CACHE_TTL,
```
Comment (abhimanyubabbar) [inline src/util/utils.js:24]: Cache till the time min(env, dnsrecordTTL) as discussed with @psrikanth88

**Review by kanishkkatara — DISMISSED** [review]

File: src/util/utils.js [hunk 2]
```diff
@@ -15,37 +16,61 @@ const BLOCK_HOST_NAMES_LIST = BLOCK_HOST_NAMES.split(',');
...
+        dnsHit = true;
+      } else {
+        const { address, ttl } = await fetchResolvedIp(hostname);
+        ip = address;
+        dnsCache.set(hostname, ip, ttl || DNS_CACHE_TTL);
+      }
+    } else {
+      const { address } = await fetchResolvedIp(hostname);
+      ip = address;
+    }
```
Comment (abhimanyubabbar) [inline src/util/utils.js:50]: What do you think if we re-write all of this with a decorator pattern ?

```javascript
@cache
function hostnameToIP(hostname) {
}
```

Inside the cache decorator we can read the env values and then fetch the IP from cache or resolve it actually.

File: src/util/utils.js [hunk 3]
```diff
@@ -15,37 +16,61 @@ const BLOCK_HOST_NAMES_LIST = BLOCK_HOST_NAMES.split(',');
...
+  // ex: [{ address: '108.157.0.0', ttl: 600 }]
+  const addresses = await resolver.resolve4(hostname, { ttl: true });
+  return addresses.length > 0 ? addresses[0] : {};
+};
 
 const staticLookup = (transformationTags) => async (hostname, _, cb) => {
-  let ips;
+  let ip;
   const resolveStartTime = new Date();
+  let dnsHit = false;
```
[the remark this answers is not in the document]
Comment (Jayachand) [reply src/util/utils.js:36]: done

File: src/util/utils.js (hunk 1 above)
Comment (Jayachand) [reply src/util/utils.js:24]: added while adding value in cache

**Review by abhimanyubabbar — APPROVED** [review]

**Review by kanishkkatara — APPROVED** [review]

**Pull request merged**